### PR TITLE
feat: expose table_properties in FFI via visit_table_properties

### PR DESCRIPTION
**Pull request by thinh2**
## What changes are proposed in this pull request?

Currently, there is no method to extract the `table_properties` from Snapshot in the FFI layer. This PR added a FFI function to get the `table_properties`. 

## How was this change tested?

A new unit test was added to verify the FFI function return the correct `table_properties`.

`cargo test --package delta_kernel_ffi --lib --all-features -- tests::test_visit_table_properties`

### CODE REVIEW

File: ffi/src/lib.rs [hunk 1]
```diff
@@ -900,6 +900,38 @@ pub unsafe extern "C" fn get_partition_columns(
     iter.into()
 }
 
+/// Visit each table property (key/value pair) for the specified snapshot by invoking the provided
+/// `visit` callback once per entry.
+///
+/// # Safety
+///
+/// Caller is responsible for passing a valid snapshot handle, and a non-null `visit` function
```
Comment (nicklan) [inline ffi/src/lib.rs]: small nit: let's just say:
```suggestion
/// Caller is responsible for passing a valid snapshot handle, and a valid `visit` function
```

It doesn't just have to be non-null, but also valid

File: ffi/src/lib.rs [hunk 2]
```diff
@@ -900,6 +900,38 @@ pub unsafe extern "C" fn get_partition_columns(
     iter.into()
 }
 
+/// Visit each table property (key/value pair) for the specified snapshot by invoking the provided
+/// `visit` callback once per entry.
+///
```
Comment (nicklan) [inline ffi/src/lib.rs:905]: Let's add:  `engine_context` is an opaque pointer passed  to each `visit` invocation.

File: ffi/src/lib.rs [hunk 3]
```diff
@@ -900,6 +900,38 @@ pub unsafe extern "C" fn get_partition_columns(
...
+/// # Safety
+///
+/// Caller is responsible for passing a valid snapshot handle, and a non-null `visit` function
+/// pointer.
+#[no_mangle]
+pub unsafe extern "C" fn visit_table_properties(
```
Comment (nicklan) [inline ffi/src/lib.rs]: I think for now let's call this `visit_table_configuration`. Within the rust kernel we parse all this and validate things, and only _then_ call it `table_properties`. 

We may want to make something more structured here in the future, but this is good for now. But let's be clear we're just looking at the config

File: ffi/src/lib.rs [hunk 4]
```diff
@@ -1099,6 +1133,63 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn test_visit_table_properties() -> Result<(), Box<dyn std::error::Error>> {
+        const METADATA_WITH_PROPS: &str = r#"{"commitInfo":{"timestamp":1587968586154,"operation":"WRITE","operationParameters":{"mode":"ErrorIfExists","partitionBy":"[]"},"isBlindAppend":true}}
```
Comment (nicklan) [inline ffi/src/lib.rs]: Let's move `METADATA_WITH_PROPS` to `test-utils` with the other similar strings

File: ffi/src/lib.rs [hunk 5]
```diff
@@ -900,6 +900,38 @@ pub unsafe extern "C" fn get_partition_columns(
...
 
+/// Visit each table properties (key/value pair) for the specified snapshot by invoking the provided
+/// `visit` callback once per entry.
+///
+/// # Safety
+///
+/// Caller is responsible for passing a valid snapshot handle, a valid `engine_context` as an opaque pointer passed to each `visit` invocation,
+/// and a valid `visit` function pointer.
+#[no_mangle]
+pub unsafe extern "C" fn visit_table_configuration(
```
Comment (chiinlquah) [inline ffi/src/lib.rs]: nit: seems like the convention is to have the snapshot before engine context 
```
  // New function (wrong order):                                                                                                                           
  pub unsafe extern "C" fn visit_table_configuration(               
      engine_context: NullableCvoid,  // <- WRONG: context before handle                                                                                   
      snapshot: Handle<SharedSnapshot>,
      visit: extern "C" fn(...),                                                                                                                           
  )                                                                 
                                                                                                                                                           
  // Convention from visit_domain_metadata, string_slice_next, etc.:                                                                                       
  pub unsafe extern "C" fn visit_domain_metadata(
      snapshot: Handle<SharedSnapshot>,  // <- handle first                                                                                                
      engine: Handle<SharedExternEngine>,                           
      engine_context: NullableCvoid,                                                                                                                       
      visitor: extern "C" fn(...),                                                                                                                         
  )   
```

File: ffi/src/lib.rs [hunk 6]
```diff
@@ -900,6 +900,38 @@ pub unsafe extern "C" fn get_partition_columns(
...
+        engine_context: NullableCvoid,
+        key: KernelStringSlice,
+        value: KernelStringSlice,
+    ),
+) {
+    let snapshot = unsafe { snapshot.as_ref() };
+    snapshot
+        .table_configuration()
+        .metadata()
+        .configuration()
```
Comment (chiinlquah) [inline ffi/src/lib.rs:924]: seems like we are only exposing metadata configuration from table config? Are we planning to expose the entire table configuration eventually? if not, I think we should make the intent clearer - visit_metadata_configuration

File: ffi/src/lib.rs (hunk 5 above)
Comment (chiinlquah) [inline ffi/src/lib.rs]: visitor? - convention

File: ffi/src/lib.rs [hunk 7]
```diff
@@ -1099,6 +1133,62 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn test_visit_table_configuration() -> Result<(), Box<dyn std::error::Error>> {
```
Comment (chiinlquah) [inline ffi/src/lib.rs]: let's also add a test on empty config

File: ffi/src/lib.rs [hunk 8]
```diff
@@ -900,6 +900,38 @@ pub unsafe extern "C" fn get_partition_columns(
     iter.into()
 }
 
+/// Visit each table properties (key/value pair) for the specified snapshot by invoking the provided
+/// `visit` callback once per entry.
+///
+/// # Safety
+///
+/// Caller is responsible for passing a valid snapshot handle, a valid `engine_context` as an opaque pointer passed to each `visit` invocation,
+/// and a valid `visit` function pointer.
```
Comment (chiinlquah) [inline ffi/src/lib.rs]: nit: previous line

File: ffi/src/lib.rs [hunk 9]
```diff
@@ -900,6 +900,38 @@ pub unsafe extern "C" fn get_partition_columns(
     iter.into()
 }
 
+/// Visit each table properties (key/value pair) for the specified snapshot by invoking the provided
```
Comment (chiinlquah) [inline ffi/src/lib.rs]: nit: grammer - property

File: ffi/src/lib.rs [hunk 10]
```diff
@@ -900,6 +900,38 @@ pub unsafe extern "C" fn get_partition_columns(
...
+        .table_configuration()
+        .metadata()
+        .configuration()
+        .iter()
+        .for_each(|(key, value)| {
+            visit(
+                engine_context,
+                kernel_string_slice!(key),
+                kernel_string_slice!(value),
+            )
```
Comment (chiinlquah) [inline ffi/src/lib.rs]: nit: Style: Missing semicolon at the end of the for_each expression statement

File: ffi/src/lib.rs [hunk 11]
```diff
@@ -1008,8 +1040,10 @@ mod tests {
     use delta_kernel::object_store::path::Path;
     use delta_kernel::object_store::ObjectStore;
     use serde_json::Value;
+    use std::collections::HashMap;
     use test_utils::{
-        actions_to_string, actions_to_string_partitioned, add_commit, TestAction, METADATA,
+        actions_to_string, actions_to_string_partitioned, actions_to_string_with_metadata,
+        add_commit, TestAction, METADATA, METADATA_WITH_CONFIGURATION,
```
Comment (chiinlquah) [inline ffi/src/lib.rs]: nit: METADATA_WITH_TABLE_PROPERTIES better?

**Review by chiinlquah — COMMENTED** [review]
Thanks for the change!! Some comments.
